pkg/store/storepb/custom_test.go

```diff
@@ -672,3 +675,135 @@ func TestSeriesRequestToPromQL(t *testing.T) {
+			inputMatchers: []LabelMatcher{
+				{Name: "__name__", Type: LabelMatcher_EQ, Value: "up"},
+			},
+			// EQ matchers are not cached
+			expectCacheAction: NoCache,
+		},
+		{
+			name: "Single RE matcher, cache miss",
+			inputMatchers: []LabelMatcher{
+				{Name: "job", Type: LabelMatcher_RE, Value: "test"},
+			},
+			expectCacheAction: CacheMiss,
+		},
+		{
+			name: "Single RE matcher, cache hit",
+			inputMatchers: []LabelMatcher{
+				{Name: "job", Type: LabelMatcher_RE, Value: "test"},
+			},
+			expectCacheAction: CacheHit,
+		},
+		{
+			name: "Multiple matchers, mixed cache",
+			inputMatchers: []LabelMatcher{
+				{Name: "__name__", Type: LabelMatcher_EQ, Value: "up"},
+				{Name: "job", Type: LabelMatcher_RE, Value: "test"},
+			},
+			// Only RE is cached, so cache hit expected for RE matcher
+			expectCacheAction: CacheHit,
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			cacheHitsBefore := getMetricValue(converter.metrics.cacheHitCount)
+			cacheTotalBefore := getMetricValue(converter.metrics.cacheTotalCount)
+
+			promMatchers, err := converter.MatchersToPromMatchers(c.inputMatchers...)
+
+			if c.expectError {
+				require.Error(t, err)
+				return
+			}
+
+			require.NoError(t, err)
+			require.Len(t, promMatchers, len(c.inputMatchers))
+
+			for i, m := range promMatchers {
+				require.Equal(t, c.inputMatchers[i].Name, m.Name)
+				require.Equal(t, c.inputMatchers[i].Value, m.Value)
+			}
+
+			if c.expectCacheAction == CacheHit {
+				require.Equal(t, cacheHitsBefore+1, getMetricValue(converter.metrics.cacheHitCount))
+				require.Equal(t, cacheTotalBefore+1, getMetricValue(converter.metrics.cacheTotalCount))
+			} else if c.expectCacheAction == CacheMiss {
+				require.Equal(t, cacheHitsBefore, getMetricValue(converter.metrics.cacheHitCount))
+				require.Equal(t, cacheTotalBefore+1, getMetricValue(converter.metrics.cacheTotalCount))
+			} else {
+				require.Equal(t, cacheHitsBefore, getMetricValue(converter.metrics.cacheHitCount))
+				require.Equal(t, cacheTotalBefore, getMetricValue(converter.metrics.cacheTotalCount))
+			}
+		})
+	}
+}
+
+func BenchmarkMatcherConverter_REWithAndWithoutCache(b *testing.B) {
+	reg := prometheus.NewRegistry()
+	converter, err := NewMatcherConverter(100, reg)
+	require.NoError(b, err)
+
+	nonTrivialRegexes := []LabelMatcher{
+		{Name: "job", Type: LabelMatcher_RE, Value: "^[a-zA-Z0-9_]+$"},
+		{Name: "env", Type: LabelMatcher_RE, Value: `prod|staging|dev`},
+		{Name: "error", Type: LabelMatcher_RE, Value: "^5\\d{2}$"},
+		{Name: "path", Type: LabelMatcher_RE, Value: "^/api/v[0-9]+/.*$"},
+		{Name: "version", Type: LabelMatcher_RE, Value: "^v[0-9]+\\.[0-9]+\\.[0-9]+$"},
+	}
+
+	b.Run("Without Cache", func(b *testing.B) {
+		for i := 0; i < b.N; i++ {
+			for _, lm := range nonTrivialRegexes {
+				_, err := MatcherToPromMatcher(lm)
+				require.NoError(b, err)
+			}
+		}
+	})
+
+	b.Run("With Cache", func(b *testing.B) {
```

Comment:
could possible to add some random cache miss cases like regex `randomkey[0-4]` and generate test data `random[0-9]` so it will have 50% hit miss